### ISSUE TITLE
Major performance issue caused by em-http-request

### ISSUE THREAD

**Issue by bawNg** [igrigorik/em-http-request#266]
I have come across what appears to be a major performance issue caused by em-http-request. After a single http request has been made using em-http-request, the EventMachine reactors tick rate appears to drop by around 86 times, forever.

This is having a serious impact on my high performance application and I will likely have to build my own minimal http client to use until this issue can be resolved.

I am sampling the EM tick rate ~940 times a second so that sampling does not cause much overhead.

```ruby
require 'em-http-request'

@tick_count = 0

def tick_rate
  (@tick_count / (EM.current_time - @tick_count_start_time)).round.tap do
    @tick_count = 0
    @tick_count_start_time = EM.current_time
  end
end

EM.epoll

EM.run do
  @tick_count_start_time = EM.current_time

  EM.add_periodic_timer(0.001) do
    @tick_count += 1
  end

  EM.add_periodic_timer(1) do
    now = Time.now
    puts "[#{now.strftime "%H:%M:%S"}] Tick rate: #{tick_rate}"
    @last_report_at = now
  end

  EM.add_timer(6) do
    http = EM::HttpRequest.new('http://google.com').get
    http.callback do
      puts "[#{Time.now.strftime "%H:%M:%S"}] API request complete"
    end

    http.errback do
      puts "[#{Time.now.strftime "%H:%M:%S"}] API request failed"
    end
  end
end
```

Output:
```
[14:49:25] Tick rate: 940
[14:49:26] Tick rate: 941
[14:49:27] Tick rate: 941
[14:49:28] Tick rate: 941
[14:49:29] Tick rate: 942
[14:49:30] Tick rate: 941
[14:49:31] API request complete
[14:49:32] Tick rate: 169
[14:49:33] Tick rate: 11
[14:49:34] Tick rate: 11
...
```

**Comment by karthikv** [comment]
I was running into this same problem; after making a single request, the event loop slowed down by a factor of 80.

After some digging, I believe the issue stems from an incorrect call to `Connection#close_connection` in the `Connection#unbind` handler at https://github.com/igrigorik/em-http-request/blob/master/lib/em-http/http_connection.rb#L198

According to the documentation at https://github.com/eventmachine/eventmachine/blob/master/lib/em/connection.rb#L177-L211, `#unbind` is called after the connection is already closed; the connection should not be closed again.

After removing this call, I no longer experienced any performance issues. I'm not familiar enough with EventMachine's C++ to know exactly why this call resulted in such a severe performance decrease.

I do, however, see signs that this diagnosis is correct. In the strace provided by @bawNg, you'll see that the tick rate slows down immediately after the following lines:

```
epoll_ctl(8, EPOLL_CTL_DEL, 9, {EPOLLIN, {u32=32629504, u64=32629504}}) = 0
epoll_ctl(8, EPOLL_CTL_DEL, 9, {EPOLLIN, {u32=32629504, u64=32629504}}) = -1 ENOENT (No such file or directory)
```

The duplicate calls to deregister the file descriptor 9 likely stem from the doubly closed connection, as per https://github.com/eventmachine/eventmachine/blob/b7f2b34d6b8d04aa8fdb017059833b64c602d819/ext/em.cpp#L759 (`ed->ShouldDelete()` is true due to the `evma_close_connection()` call).

Let me know if my logic is incorrect or if I missed something.

**Comment by igrigorik** [comment]
Interesting, thanks for tracking this down! 

I don't think #279 is sufficient / correct solution. Unbind is also aliased to `close` which is the client API for terminating a connection, hence we can't just remove the connection_close call. I'm wondering if this should be solved in EM core: em-http won't be the only one hitting this gotcha... and it's definitely a costly performance gotcha. @tmm1 thoughts?

**Comment by bawNg** [comment]
Nice catch karthikv. I had just finished benchmarking this issue again, this time on Windows, and was about to start investigating further.

Ideally this should be fixed in the EM core, but since EM is maintained very infrequently, a workaround implemented by em-http would be likely help a lot short term.

Since `unbind` is aliased to `close`, the connection state will need to be tracked in `connection_completed` and `unbind` so that `close_connection` is only called when a connection is still active.

**Comment by bawNg** [comment]
I can't push a patch to github easily right now, but here is a proof of concept monkeypatch which solves the performance issues.

```ruby
module EventMachine
  class HttpStubConnection
    alias_method :_connection_completed, :connection_completed
    alias_method :_unbind, :unbind
    alias_method :_close_connection, :close_connection

    def connection_completed
      @is_connected = true
      _connection_completed
    end

    def unbind(reason=nil)
      @is_connected = false
      _unbind(reason)
    end

    def close_connection
      _close_connection if @is_connected
    end
  end
end
```

**Comment by karthikv** [comment]
Ah, I didn't realize `unbind` was an alias to `close`. I've updated the PR based on @bawNg's suggestion. Let me know if there's anything else I need to fix.

All the tests pass, except 'should fail GET on DNS timeout.' This was failing before I introduced any changes, so I doubt it's affected by the PR.

**Comment by igrigorik** [comment]
@bawNg can you confirm that this patch addresses your issue as well?

I'm not against merging a fix into em-http, but I'd *much* prefer this to be sent upstream into EM core.. This seems like a really costly bug that many others could be (unknowingly) suffering from as well. Would love to hear @tmm1's input.

**Comment by bawNg** [comment]
The patch looks good, it does exactly the same thing as my monkeypatch.

I agree this should be done by the EM core so that users don't need to account for it on the application level. It is a very obscure issue with drastic effects and I'm surprised no one else spotted the performance issues sooner.

I'd like to see simple tick rate profiling implemented by the EM core in the future, so that it is possible to profile an accurate number of ticks per second efficiently instead of sampling using timers. I'm currently sampling the tick rate using a self-calling once off 0.01 timer to sync with the 16ms default EM timer frequency, which provides ~62 ticks per second on all hardware.

**Comment by karthikv** [comment]
This was actually solved in EventMachine core 1.0.4 update. See eventmachine/eventmachine#441 and eventmachine/eventmachine#445. I tested my own code and @bawNg's script in #266; both worked without reducing the tick rate in EventMachine 1.0.4 and later.
